### PR TITLE
feat(cast) Add bit shifting utilities

### PR DESCRIPTION
**Pull request by iFrostizz**
## Motivation

Implementing https://github.com/foundry-rs/foundry/issues/2394
This new feature will allow to perform bitwise shifting operations from cast.

### CODE REVIEW

**Comment by iFrostizz** [comment]
I'm quite septic about the command used here, adding a `<<` or `>>` command as discussed in the issue seems to make the usage of quotes obligatory.

Also, usually values are coming after argument names with CLI apps, and it's the case if I follow the mainline of cast.

So the command would look like:

`cast '>>' 4 1` for "right shift this decimal number 4 by one byte", and this is not what @HrikB wanted.

Does that makes sense, or should we rename the command / make a custom command to fit what was wanted as start ?

File: cli/src/opts/cast.rs
Comment (onbjerg) [inline cli/src/opts/cast.rs:145]: As discussed in the issue the UX should be `cast shl <shift> <value>` (similar for right shift using `shr`). Would also discourage using `ls` as an alias - and generally I don't think we need any aliases for these when the names are `shl` and `shr`

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1072]: Imo we should try to be smarter about this (e.g. if no `0x` but the string contains a-f then we assume hex) since other commands support hex with and without prefix.

Also noting here that the output base should be configurable. In the issue the flag `--output` was suggested but I think a better name is probably `--base-out <base>`. Should only support base 10 and base 16 (maybe using the values `10`/`16` and `dec`/`hex`). The output base should default to hex as this is default for all other commands as well.

Comment (iFrostizz) [reply cast/src/lib.rs:1072]: For the input base detection, would you recommend the usage of regex ?
Also, for the output should we suffix by 0x if it's an hex ?

Comment (onbjerg) [reply cast/src/lib.rs:1072]: I'd check the other commands, I don't remember if `0x` is outputted or not (I don't think it is)

For input base detection you don't need regex, just:

- Strip `0x` if it's there. If it is, assume hex
- If it's not, try converting from base 10, I think, and if that fails, try converting from base 16
- If above step fails, fail

We could also add a `--base-in <base>` flag for maximum control in the case someone passes e.g. `90` but want's it interpreted as hex (this is still possible by prepending `0x` but might be too obscure?)

Comment (iFrostizz) [reply cast/src/lib.rs:1072]: I found some places where it was indeed outputted with a trailing "0x":

- https://github.com/foundry-rs/foundry/blob/master/cast/src/lib.rs#L103-L104
- https://github.com/foundry-rs/foundry/blob/master/cast/src/lib.rs#L982-L991
- https://github.com/foundry-rs/foundry/blob/master/cast/src/tx.rs#L234

On the other side, the U256 type `from_str` trait appear to assume that the base is always 16: https://github.com/paritytech/parity-common/blob/master/ethereum-types/src/uint.rs#L68

So I think that a --base-in would be useful, I hope that it wouldn't be too verbose. Now, we have two options:

- Make the --base-in require and there won't be any confusion, but could be annoying to type
- Don't make it required and throw an error that explains that the `--base-in` argument should be added in order to be more explicit if it could be both an hex or dec number, and that no "0x" has been prepended.

Comment (onbjerg) [reply cast/src/lib.rs:1072]: Yeah I think we should make `--base-in` optional

File: cli/src/opts/cast.rs
Comment (iFrostizz) [reply cli/src/opts/cast.rs:145]: Wait, would that be `cast shl <shift> <value>` or `cast shl <value> <shift>` as my example ?

**Comment by iFrostizz** [comment]
Seems to have got something working, I still need to pretty print the hex, because currently it looks like this:

```console
$ tcast shr 4 1 --base-out 16
0x0000000000000000000000000000000000000000000000000000000000000002
```
I haven't noticed (probably missed) a function to remove those trailing 0, so maybe we could run a for loop until a value != 0 is reached.

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:736]:
```suggestion
            _ => eyre::bail!("Unknown input base: {base_in}"),
```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:740]: don't need this here since we will assume hex because of 0x (we handle the error in the command itself instead)

```suggestion

```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:739]: you can change this to

```suggestion
        None if value.starts_with("0x") => {
```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:743]: Using the other pattern change, we can change this to

```suggestion
        None => {
```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:747]: i'm not sure about this logic - you're first checking if it can be decoded as base 10, if it can, then u check if it can be decoded as base 16?

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:771]: should work, `U256` impls `LowerHex`

```suggestion
        16 => Ok(format!("0x{:x}", val)),
```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:772]:
```suggestion
        _ => Err(eyre::eyre!("Unknown output base: {base_out}")),
```

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1060]: Use the results instead. You can provide better context to the user with `wrap_err`: https://docs.rs/eyre/latest/eyre/trait.WrapErr.html#tymethod.wrap_err

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1067]: Use the results here as well

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1058]:
```suggestion

```

**Review by onbjerg — COMMENTED** [review]
Looking good, some comments. In general don't use `expect` or `unwrap` if you're dealing with user input unless you know 100% that it should not be able to panic. Our panic handler tells the user that they encountered a bug, but our error handler tells the user that something with their input is wrong

File: cli/src/cast.rs
Comment (iFrostizz) [reply cli/src/cast.rs:747]: An hex number can always be parsed as a dec number but a dec cannot because it can contains letters.

So first check:
- Can it be parsed as a decimal number ?
    - Yes => Can it be parsed as hex ?
        - Yes => Then the base is ambiguous as the 0x is not here, it can either be hex or dec
        - No => It must be a dec
    - No => It may be an hex

Comment (onbjerg) [reply cli/src/cast.rs]: Yes, but this logic is more like

- Can it be parsed as decimal?
  - Yes: Can it be parsed as hex?
    - No: Fail
    - Yes: Assume decimal
  - No: It may be hex

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1059]:
```suggestion
            .wrap_err("Cannot parse input as base {base_in}")?;
```

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1060]: Do we want the shift to be hex/dec too? Afaik you mostly just do shifts in dec

File: cast/src/lib.rs
Comment (onbjerg) [inline cast/src/lib.rs:1068]:
```suggestion
            .wrap_err("Cannot parse input with as base {base_in}")?;
```

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:742]: If the value can be decoded as base 10, then it can always be decoded as base 16, so this check is redundant.

However, this also does not solve the original issue we tried to solve: if it can be decoded as base 10 **and** it can be decoded as base 16, then it is ambigious, and that is the point we should error, but we just assume base 10. Does that make sense?

Also the error should be something like:

"Could not autodetect input base: input could be decimal or hexadecimal. Please prepend with 0x if the input is hexadecimal."

File: cli/src/cast.rs
Comment (onbjerg) [inline cli/src/cast.rs:747]:
```suggestion
                U256::from_str_radix(value, 16).expect("Could not autodetect input base.");
```

File: cast/src/lib.rs
Comment (iFrostizz) [reply cast/src/lib.rs:1060]: Yes, that's stupid you're right !

File: cli/src/cast.rs
Comment (iFrostizz) [reply cli/src/cast.rs:742]: IMO, we should be more explicit than this message, and add a ", or specify --base-in parameter" because the user may get confused if he is just trying to shift a decimal number.

Comment (iFrostizz) [reply cli/src/cast.rs:742]: Also, I've seen in other cast utilities commands a panic in the case when an user input could not be parsed because the program would panic, and display this:

```
The application panicked (crashed).
Message:  Could not autodetect input base.: FromStrRadixErr { kind: InvalidCharacter, source: Some(Hex(FromHexError { inner: InvalidHexCharacter { c: 't', index: 2 } })) }
```

Should we refactor all places where the code could panic because of an `unwrap()` or `expect()` after parsing of the user input in order to just send the friendly message, or is that fine that way ?

File: cli/src/cast.rs
Comment (mattsse) [inline cli/src/cast.rs:754]: this will match anything due to the `_` wildcard pattern, is this intended here?

File: cli/src/opts/cast.rs
Comment (mattsse) [inline cli/src/opts/cast.rs:879]: same

**Review by mattsse — COMMENTED** [review]
question re "16" base

otherwise code lgtm, defer to @onbjerg for final approval

File: cli/src/cast.rs
Comment (iFrostizz) [reply cli/src/cast.rs:754]: Right, so the aim was to default the base out to 16 but we might prefere to eyre::bail! a message for the user here

File: cli/src/cast.rs
Comment (onbjerg) [reply cli/src/cast.rs]: Yes, but that is part of #2131. Only making you do it for this PR so we have less work once we get around to it :)

File: cli/src/cast.rs
Comment (onbjerg) [reply cli/src/cast.rs:754]: I'd throw an error to the user in this case

**Comment by onbjerg** [comment]
LGTM once @mattsse comment is addressed + lint passes

File: cli/src/cast.rs
Comment (iFrostizz) [reply cli/src/cast.rs]: Lemme know if I can help

**Review by mattsse — APPROVED** [review]

**Pull request merged**
